services/src/modules/directives/error-handler.ts

```diff
@@ -0,0 +1,125 @@
+import { gql } from 'apollo-server-core';
+import { GraphQLFieldResolverParams } from 'apollo-server-types';
+import { defaultFieldResolver, FieldNode, GraphQLField, GraphQLObjectType, GraphQLResolveInfo } from 'graphql';
+import { SchemaDirectiveVisitor } from 'graphql-tools';
+import { injectArgs } from '../arguments-injection';
+import { RequestContext } from '../context';
+
+const defaultObjectResolver = (
+  source: unknown,
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  _fields: Record<string, ReadonlyArray<FieldNode>>,
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  _context: RequestContext,
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  _info: GraphQLResolveInfo
+) => source;
+
+export class ErrorHandlerDirective extends SchemaDirectiveVisitor {
+  visitObject(object: GraphQLObjectType<unknown, RequestContext>) {
+    const originalResolve = object.resolveObject || defaultObjectResolver;
+    const { catchError, throwError } = this.args as ErrorHandlerDirectiveArgs;
+
+    object.resolveObject = async (source, fields, context, info) => {
+      const injectionArgs = { source, args: {}, context, info };
+      let result;
+      try {
+        result = await originalResolve.call(object, source, fields, context, info);
+      } catch (error) {
+        result = handleCatchError(catchError, error, injectionArgs, result);
+      }
+      throwError && handleThrowError(throwError, result, injectionArgs);
+      return result;
+    };
+  }
+
+  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
+    const originalResolve = field.resolve || defaultFieldResolver;
+    const { catchError, throwError } = this.args as ErrorHandlerDirectiveArgs;
+
+    field.resolve = async (source, args, context, info) => {
+      const injectionArgs = { source, args, context, info };
+      let result;
+      try {
+        result = await originalResolve.call(field, source, args, context, info);
+      } catch (error) {
+        result = handleCatchError(catchError, error, injectionArgs, result);
+      }
+
+      throwError && handleThrowError(throwError, result, injectionArgs);
+      return result;
+    };
+  }
+}
+
+export const sdl = gql`
+  input CatchErrorInput {
+    condition: String
+    returnValue: JSON
+  }
+
+  input ThrowErrorInput {
+    condition: String
+    errorToThrow: String
+  }
+
+  directive @errorHandler(catchError: CatchErrorInput, throwError: ThrowErrorInput) on OBJECT | FIELD_DEFINITION
+`;
+
+type ErrorHandlerDirectiveArgs = {
+  catchError?: {
+    condition?: string;
+    returnValue: string | Record<string, unknown>;
+  };
+  throwError?: {
+    condition?: string;
+    errorToThrow: string;
+  };
+};
+
+function handleCatchError(
+  catchError: { condition?: string | undefined; returnValue?: string | Record<string, unknown> } | undefined,
+  error: Error,
+  injectionArgs: GraphQLFieldResolverParams<unknown, RequestContext>,
+  result: unknown
```

Comment:
The function shouldn't get this parameter (and doesn't use its value), should just be defined as a variable inside the function